contracts/external/oracles/ChainlinkAutomationPriceOracle.sol

```diff
@@ -0,0 +1,179 @@
+        Require.that(
+            tx.origin == address(0),
+            _FILE,
+            "static rpc calls only"
+        );
+
+        return (_checkUpkeepConditions(), "0x");
+    }
+
+    function performUpkeep(bytes calldata /* performData */) external {
+        Require.that(
+            msg.sender == CHAINLINK_REGISTRY,
+            _FILE,
+            "caller is not Chainlink"
+        );
+        Require.that(
+            _checkUpkeepConditions(),
+            _FILE,
+            "checkUpkeep conditions not met"
+        );
+
+        _updateExchangeRateAndTimestamp();
+    }
+
+    // ============================ Internal Functions ============================
+
+    function _ownerSetHeartbeat(uint256 _heartbeat) internal {
+        emit HeartbeatSet(_heartbeat);
+        HEARTBEAT = _heartbeat;
+    }
+
+    function _ownerSetGracePeriod(uint256 _gracePeriod) internal {
+        emit GracePeriodSet(_gracePeriod);
+        GRACE_PERIOD = _gracePeriod;
+    }
+
+    function _ownerSetUpperEdge(uint256 _upperEdge) internal {
+        Require.that(
+            _upperEdge > 10000,
+            _FILE,
+            "Invalid upper edge"
+        );
+        emit UpperEdgeSet(_upperEdge);
+        UPPER_EDGE = _upperEdge;
+    }
+
+    function _ownerSetLowerEdge(uint256 _lowerEdge) internal {
+        Require.that(
+            _lowerEdge < 10000,
+            _FILE,
+            "Invalid lower edge"
+        );
+        emit LowerEdgeSet(_lowerEdge);
+        LOWER_EDGE = _lowerEdge;
+    }
+
+    function _ownerSetChainlinkRegistry(address _chainlinkRegistry) internal {
+        Require.that(
+            _chainlinkRegistry != address(0),
+            _FILE,
+            "Invalid chainlink registry"
+        );
+        emit ChainlinkRegistrySet(_chainlinkRegistry);
+        CHAINLINK_REGISTRY = _chainlinkRegistry;
+    }
+
+    function _updateExchangeRateAndTimestamp() internal {
+        (exchangeRateNumerator, exchangeRateDenominator) = _getExchangeRate();
+        latestTimestamp = block.timestamp;
+    }
+
+    function _checkUpkeepConditions() internal view returns (bool) {
+        (uint256 currentNumerator, uint256 currentDenominator) = _getExchangeRate();
+        if (currentDenominator == 0) { return false; }
+
+        uint256 cachedExchangeRate = exchangeRateNumerator * _ONE_UNIT / exchangeRateDenominator;
+        uint256 currentExchangeRate = currentNumerator * _ONE_UNIT / currentDenominator;
+
+        uint256 upperEdge = cachedExchangeRate * UPPER_EDGE / 10000;
+        uint256 lowerEdge = cachedExchangeRate * LOWER_EDGE / 10000;
+        return (
+            currentExchangeRate >= upperEdge ||
+            currentExchangeRate <= lowerEdge ||
+            block.timestamp >= latestTimestamp + HEARTBEAT
+        );
+    }
+
+    // ============================ Virtual Functions ============================
+
+    function _getExchangeRate() internal virtual view returns (uint256, uint256);
```

Comment:
Make the return variables named here so it's clearer what each `uint256` signifies